dashboard/app/models/pd/payment/payment_calculator_base.rb

```diff
@@ -0,0 +1,164 @@
+    # @return [Hash{String => Numeric}] Map of payment parts (e.g. :food) to their dollar amounts.
+    def calculate_payment_amounts(payment)
+      {}
+    end
+
+    # Gets session attendance summaries for a workshop.
+    # @param workshop [Pd::Workshop]
+    # @return [Array<SessionAttendanceSummary>] summary of attendance for each session.
+    def get_session_attendance_summaries(workshop)
+      workshop.sessions.map do |session|
+        teacher_ids = Pd::Attendance.where(pd_session_id: session.id).pluck(:teacher_id)
+        SessionAttendanceSummary.new session.hours, teacher_ids
+      end
+    end
+
+    private
+
+    # Calculates raw attendance totals per teacher based on a supplied set of session attendance summaries.
+    #
+    # @param session_attendance_summaries [Array<SessionAttendanceSummary>]
+    # summary of teacher attendance for each session
+    #
+    # @return [Hash{Integer => TeacherAttendanceTotal}] map of teacher id to raw attendance totals for that teacher.
+    def calculate_raw_teacher_attendance(session_attendance_summaries)
+      {}.tap do |raw_attendance_per_teacher|
+        session_attendance_summaries.each do |session_attendance_summary|
+          hours = session_attendance_summary.hours
+          session_attendance_summary.teacher_ids.each do |teacher_id|
+            if raw_attendance_per_teacher[teacher_id]
+              raw_attendance_per_teacher[teacher_id].add_session hours
+            else
+              raw_attendance_per_teacher[teacher_id] = TeacherAttendanceTotal.new 1, hours
+            end
+          end
+        end
+      end
+    end
+
+    # Calculates adjusted teacher attendance, after applying min / max attendance and qualification rules
+    # Filters out unqualified teachers, and those with 0 attendance.
+    # Note - calls #teacher_qualified? for each teacher id
+    #
+    # @param raw_teacher_attendance [Hash{Integer => TeacherAttendanceTotal}]
+    #   map of teacher id to raw attendance total for that teacher.
+    # @param min_attendance_days [Integer] minimum days a teacher must be in attendance to count.
+    # @param num_days [Integer] max number of days that will count for this workshop
+    # @param num_hours [Integer] max number of hours that will count for this workshop
+    #
+    # @return [Hash{Integer => TeacherAttendanceTotal}] map of teacher id to adjusted attendance totals for that teacher.
+    def adjust_teacher_attendance(raw_teacher_attendance, min_attendance_days, num_days, num_hours)
+      {}.tap do |adjusted_attendance_per_teacher|
+        # Require min attendance, filter out unqualified teachers, and apply caps (via @num_days & @num_hours).
+        raw_teacher_attendance.each do |teacher_id, teacher_attendance|
+          adjusted_days = teacher_attendance.days < min_attendance_days ? 0 : [teacher_attendance.days, num_days].min
+          next unless adjusted_days > 0 && teacher_qualified?(teacher_id)
```

Comment:
I'm inclined to do the teacher_qualified? on line 154 - raw_teacher_attendendence.keep_if etc